# NB05: PGLS Regression — Does Metal Gene Cluster Burden Predict Niche Breadth?

**Question**: After controlling for shared evolutionary history (phylogeny), does metal resistance
gene burden predict ecological niche breadth in bacteria?

**Method**: Phylogenetic Generalized Least Squares (PGLS) via `ape::gls + nlme::corPagel`.  
Pagel's λ is estimated jointly with the regression coefficients (ML), so residual covariance
is neither fully phylogenetic (λ=1) nor fully independent (λ=0) — the data determine the blend.

**Models**:
```
mean_levins_B_std  ~  mean_n_metal_clusters_z     (simple × 3 predictors)
mean_levins_B_std  ~  mean_metal_core_fraction_z
mean_levins_B_std  ~  mean_n_metal_types_z
mean_n_envs        ~  mean_n_metal_clusters_z     (same 3 predictors)
mean_n_envs        ~  mean_metal_core_fraction_z
mean_n_envs        ~  mean_n_metal_types_z
mean_levins_B_std  ~  all 3 metal gene predictors jointly     (multiple predictor)
```
Predictors are z-scored so β coefficients are directly comparable.

**Inputs**:
- `data/gtdb_bac_genus_pruned.tree` — 2,286-tip bacterial genus tree (NB04)
- `data/genus_trait_table.csv` — niche breadth + metal gene clusters per genus (NB04)

**Outputs**:
- `data/pgls_results.csv` — β, SE, t, p, λ, ΔAIC per simple model
- `data/pgls_multi_results.csv` — multi-predictor model coefficients
- `figures/pgls_forest_plot.png` — coefficient forest plot
- `figures/pgls_scatter.png` — partial regression scatter plots

**Analytical subset**: 606 bacterial genera with n_otus ≥ 3, in GTDB tree, and metal gene cluster data.

**Bonferroni threshold** (6 simple models): p < 0.0083

In [1]:
import subprocess
import sys
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns
from scipy.optimize import minimize_scalar
from scipy import stats as sp_stats
from Bio import Phylo

DATA    = Path('../data')
FIGURES = Path('../figures')
SCRIPTS = Path('../scripts')
FIGURES.mkdir(exist_ok=True)

print('Paths ready.')
print(f'  DATA:    {DATA.resolve()}')
print(f'  FIGURES: {FIGURES.resolve()}')

Paths ready.
  DATA:    /home/hmacgregor/BERIL-research-observatory/projects/microbeatlas_metal_ecology/data
  FIGURES: /home/hmacgregor/BERIL-research-observatory/projects/microbeatlas_metal_ecology/figures


## 1. Load data and prepare PGLS subset

In [2]:
traits = pd.read_csv(DATA / 'genus_trait_table.csv')
print(f'Trait table: {len(traits):,} genera × {len(traits.columns)} columns')

AMR_COLS = ['mean_n_metal_clusters', 'mean_metal_core_fraction', 'mean_n_metal_types']
RESPONSE_COLS = ['mean_levins_B_std', 'mean_n_envs']

# Verify columns present
for col in AMR_COLS + RESPONSE_COLS:
    present = col in traits.columns
    print(f'  {col:<35s}  {"OK" if present else "MISSING"}')

# PGLS subset: ≥3 OTUs, in bacterial tree, has metal cluster data + both response variables
bac_tree_tips = {t.name for t in Phylo.read(str(DATA / 'gtdb_bac_genus_pruned.tree'), 'newick').get_terminals()}

pgls_mask = (
    (traits['n_otus'] >= 3) &
    traits['genus_lower'].isin(bac_tree_tips) &
    traits['mean_n_metal_clusters'].notna() &
    traits['mean_levins_B_std'].notna() &
    traits['mean_n_envs'].notna()
)
pgls_df = traits[pgls_mask].copy()
print(f'\nPGLS subset: {len(pgls_df):,} genera')
print(f'  Nitrifier genera in subset: {pgls_df["is_nitrifier"].sum()}')
print(f'  Phyla represented: {pgls_df["phylum"].nunique()}')

# Z-score predictors
for col in AMR_COLS:
    pgls_df[col + '_z'] = (pgls_df[col] - pgls_df[col].mean()) / pgls_df[col].std()
    print(f'  {col}: mean={pgls_df[col].mean():.3f}  sd={pgls_df[col].std():.3f}')

# Save filtered subset for R
pgls_df.to_csv(DATA / 'pgls_subset.csv', index=False)
print(f'\nPGLS subset saved to data/pgls_subset.csv')

Trait table: 2,851 genera × 22 columns
  mean_n_metal_clusters                OK
  mean_metal_core_fraction             OK
  mean_n_metal_types                   OK
  mean_levins_B_std                    OK
  mean_n_envs                          OK

PGLS subset: 1,000 genera
  Nitrifier genera in subset: 3
  Phyla represented: 29
  mean_n_metal_clusters: mean=13.828  sd=8.638
  mean_metal_core_fraction: mean=0.716  sd=0.186
  mean_n_metal_types: mean=4.989  sd=1.670

PGLS subset saved to data/pgls_subset.csv


## 2. PGLS via R (ape + nlme)

`ape::gls` with `nlme::corPagel` correlation structure.  
The R script estimates Pagel's λ jointly with β via ML, so the residual covariance
reflects the fraction of variance that is phylogenetically structured.

```r
fit <- gls(mean_levins_B_std ~ mean_n_metal_clusters_z,
           data       = sub,
           correlation = corPagel(0.5, phy=tree, fixed=FALSE),
           method      = 'ML')
```

In [3]:
import shutil

# Check R and required packages
check = subprocess.run(
    ['/home/hmacgregor/r_env/bin/Rscript', '-e',
     'for (pkg in c("ape","nlme")) { ok <- requireNamespace(pkg, quietly=TRUE); '
     'cat(pkg, if(ok) "OK" else "MISSING", "\\n") }'],
    capture_output=True, text=True
)
print('R package check:')
print(check.stdout.strip())

r_script = SCRIPTS / 'pgls_regression.R'
print(f'\nR script: {r_script}  ({sum(1 for _ in r_script.read_text().splitlines())} lines)')

R package check:
ape OK 
nlme OK

R script: ../scripts/pgls_regression.R  (240 lines)


In [4]:
# Run the R PGLS script
# Runtime: typically < 2 min for 606 genera × 7 models
data_dir = str(DATA.resolve())
print(f'Running pgls_regression.R (data_dir={data_dir}) ...')

result = subprocess.run(
    ['/home/hmacgregor/r_env/bin/Rscript', str(r_script.resolve()), data_dir],
    capture_output=True, text=True, timeout=600
)
print(result.stdout)
if result.returncode != 0:
    print('--- STDERR ---')
    print(result.stderr[-2000:])

Running pgls_regression.R (data_dir=/home/hmacgregor/BERIL-research-observatory/projects/microbeatlas_metal_ecology/data) ...



=== PGLS regression (ape + nlme) ===
Data dir: /home/hmacgregor/BERIL-research-observatory/projects/microbeatlas_metal_ecology/data

Tree:        2283 tips
Trait table: 2851 genera, 22 cols

PGLS subset:  1000 genera (n_otus>=3, in tree, has AMR + niche breadth)
  Predictor means and SDs (raw):
    mean_n_metal_clusters                mean=13.828  sd=8.638
    mean_metal_core_fraction             mean=0.716  sd=0.186
    mean_n_metal_types                   mean=4.989  sd=1.670

Pruned tree:  1000 tips

=== Simple PGLS models ===

[PGLS] mean_levins_B_std ~ mean_n_metal_clusters_z  (n=1000)
  lambda=0.8803  beta=0.0087  SE=0.0036  t=2.442  p=0.01479  deltaAIC=-3.78

[PGLS] mean_levins_B_std ~ mean_metal_core_fraction_z  (n=1000)
  lambda=0.8906  beta=0.0036  SE=0.0031  t=1.155  p=0.2484  deltaAIC=0.67

[PGLS] mean_levins_B_std ~ mean_n_metal_types_z  (n=1000)
  lambda=0.8862  beta=0.0139  SE=0.0055  t=2.497  p=0.01268  deltaAIC=-4.19

[PGLS] mean_n_envs ~ mean_n_metal_clusters_z  (n=1

## 2b. Python PGLS fallback (scipy + BioPython)

Runs if R is unavailable. Uses the same profile likelihood approach as NB04:
- Build VCV matrix C from the pruned tree
- λ-transform: C_λ[i,j] = λ×C[i,j] (i≠j); C_λ[i,i] = C[i,i]
- Profile logL over λ with β and σ² analytically maximised (GLS)
- At λ_ML: fit GLS to get β, SE, t, p

In [5]:
def tree_vcv_fast(tree_path, tip_names):
    """
    Build phylogenetic VCV matrix C from a pruned newick tree.
    C[i,i] = root-to-tip distance; C[i,j] = root-to-MRCA distance.

    Uses a single DFS to compute root-to-node depths, then n²/2 MRCA lookups
    (faster than n² pairwise tree.distance() calls).
    """
    tree = Phylo.read(str(tree_path), 'newick')
    tip_map = {t.name: t for t in tree.get_terminals()}
    n = len(tip_names)

    # DFS: compute root-to-every-node depth
    node_depth = {}  # id(clade) → distance from root
    def dfs(clade, d):
        node_depth[id(clade)] = d
        for child in clade.clades:
            dfs(child, d + (child.branch_length or 0.0))
    dfs(tree.root, 0.0)

    C = np.zeros((n, n))
    for i, name in enumerate(tip_names):
        C[i, i] = node_depth[id(tip_map[name])]

    for i in range(n):
        for j in range(i + 1, n):
            mrca = tree.common_ancestor([tip_map[tip_names[i]], tip_map[tip_names[j]]])
            C[i, j] = C[j, i] = node_depth[id(mrca)]
    return C


def pgls_profile_logL(lam, C, Y, X):
    """Profile log-likelihood for PGLS with λ-transformed VCV."""
    n, p = X.shape
    C_lam = lam * C.copy()
    np.fill_diagonal(C_lam, np.diag(C))
    C_lam += np.eye(n) * 1e-10
    try:
        iC = np.linalg.inv(C_lam)
    except np.linalg.LinAlgError:
        return -np.inf
    XtiCX = X.T @ iC @ X
    try:
        beta = np.linalg.solve(XtiCX, X.T @ iC @ Y)
    except np.linalg.LinAlgError:
        return -np.inf
    resid = Y - X @ beta
    RSS = resid @ iC @ resid
    sigma2 = RSS / n
    if sigma2 <= 0:
        return -np.inf
    sign, logdet = np.linalg.slogdet(C_lam)
    if sign <= 0:
        return -np.inf
    return -n / 2 * np.log(2 * np.pi) - n / 2 * np.log(sigma2) - 0.5 * logdet - n / 2


def pgls_gls(C_lam, Y, X):
    """GLS with given (λ-transformed) VCV matrix. Returns β, SE, t, p."""
    n, p = X.shape
    iC = np.linalg.inv(C_lam)
    XtiCX = X.T @ iC @ X
    beta = np.linalg.solve(XtiCX, X.T @ iC @ Y)
    resid = Y - X @ beta
    RSS = resid @ iC @ resid
    sigma2 = RSS / (n - p)
    var_beta = sigma2 * np.linalg.inv(XtiCX)
    SE = np.sqrt(np.diag(var_beta))
    t_stat = beta / SE
    p_val = 2 * sp_stats.t.sf(np.abs(t_stat), df=n - p)
    return beta, SE, t_stat, p_val


def run_pgls_python(tree_path, traits_df, response_col, predictor_cols,
                    label='', verbose=True):
    """
    Run PGLS for one model: response ~ predictors (all z-scored in traits_df).
    Returns a dict of results or None.
    """
    tree = Phylo.read(str(tree_path), 'newick')
    tip_set = {t.name for t in tree.get_terminals()}

    all_cols = [response_col] + predictor_cols
    sub = traits_df[traits_df['genus_lower'].isin(tip_set)].dropna(subset=all_cols).copy()
    tip_names = list(sub['genus_lower'])
    Y = sub[response_col].values.astype(float)
    X = np.column_stack([np.ones(len(tip_names))] +
                        [sub[pc].values.astype(float) for pc in predictor_cols])

    if verbose:
        print(f'  [{label}] {response_col} ~ {" + ".join(predictor_cols)}: {len(Y)} genera')
    if len(Y) < 20:
        if verbose: print('  SKIP: < 20 genera')
        return None

    C = tree_vcv_fast(tree_path, tip_names)

    opt = minimize_scalar(
        lambda lam: -pgls_profile_logL(lam, C, Y, X),
        bounds=(0.0, 1.0), method='bounded', options={'xatol': 1e-6}
    )
    lam_ml = opt.x
    logL_ml = -opt.fun

    C_lam = lam_ml * C.copy()
    np.fill_diagonal(C_lam, np.diag(C))
    C_lam += np.eye(len(Y)) * 1e-10
    beta, SE, t_stat, p_val = pgls_gls(C_lam, Y, X)

    # Intercept-only logL
    X0 = np.ones((len(Y), 1))
    logL_0 = pgls_profile_logL(lam_ml, C, Y, X0)
    delta_AIC = 2 * (len(predictor_cols)) - 2 * (logL_ml - logL_0)

    rows = []
    for i, pc in enumerate(predictor_cols):
        sig = ('***' if p_val[i+1] < 0.001 else
               ('**' if p_val[i+1] < 0.01 else
               ('*' if p_val[i+1] < 0.05 else 'ns')))
        if verbose:
            print(f'    {pc:<40s}  β={beta[i+1]:+.4f}  p={p_val[i+1]:.4g}  {sig}')
        rows.append({
            'response': response_col, 'predictor': pc, 'n_taxa': len(Y),
            'lambda': lam_ml, 'beta': beta[i+1], 'SE': SE[i+1],
            't_stat': t_stat[i+1], 'p_value': p_val[i+1],
            'delta_AIC': delta_AIC, 'logL': logL_ml,
        })
    return rows


print('Python PGLS functions defined.')

Python PGLS functions defined.


In [6]:
# Run Python PGLS if R results not yet available
results_path = DATA / 'pgls_results.csv'

if results_path.exists():
    print('pgls_results.csv found (from R run) — Python fallback not needed.')
    print('Python PGLS will run anyway on the multi-predictor model as verification.')
else:
    print('pgls_results.csv not found — running Python PGLS ...')
    pgls_traits = pd.read_csv(DATA / 'pgls_subset.csv')
    bac_pruned = DATA / 'gtdb_bac_genus_pruned.tree'

    AMR_Z = [c + '_z' for c in AMR_COLS]
    all_rows = []
    for resp in RESPONSE_COLS:
        for pred_z in AMR_Z:
            rows = run_pgls_python(bac_pruned, pgls_traits, resp, [pred_z], label='Bacteria')
            if rows:
                all_rows.extend(rows)

    if all_rows:
        py_results = pd.DataFrame(all_rows)
        py_results.to_csv(results_path, index=False)
        print(f'\nPython PGLS results saved ({len(py_results)} models)')
        print(py_results[['response', 'predictor', 'n_taxa', 'lambda', 'beta', 'p_value']].to_string(index=False))

print('Done.')

pgls_results.csv found (from R run) — Python fallback not needed.
Python PGLS will run anyway on the multi-predictor model as verification.
Done.


## 3. Results: simple PGLS models

In [7]:
results = pd.read_csv(DATA / 'pgls_results.csv')

# Clean predictor labels for display
pred_labels = {
    'mean_n_metal_clusters_z':     'Metal gene clusters (z)',
    'mean_metal_core_fraction_z':  'Core metal fraction (z)',
    'mean_n_metal_types_z':        'Metal types (z)',
}
resp_labels = {
    'mean_levins_B_std': "Levins' B_std",
    'mean_n_envs':       'Environments detected',
}
BONFERRONI = 0.05 / 6   # 6 simple models

results['predictor_label'] = results['predictor'].map(pred_labels).fillna(results['predictor'])
results['response_label']  = results['response'].map(resp_labels).fillna(results['response'])
results['sig'] = results['p_value'].apply(
    lambda p: '***' if p < 0.001 else ('**' if p < 0.01 else ('*' if p < 0.05 else 'ns')))
results['bonf_sig'] = results['p_value'] < BONFERRONI

display_cols = ['response_label', 'predictor_label', 'n_taxa', 'lambda',
                'beta', 'SE', 't_stat', 'p_value', 'delta_AIC', 'sig']

print(f'Simple PGLS results (Bonferroni threshold: p < {BONFERRONI:.4f}):')
print()
for resp in RESPONSE_COLS:
    sub = results[results['response'] == resp]
    print(f'Response: {resp_labels[resp]}')
    print(sub[display_cols].to_string(index=False))
    print()

Simple PGLS results (Bonferroni threshold: p < 0.0083):

Response: Levins' B_std
response_label         predictor_label  n_taxa   lambda     beta       SE   t_stat  p_value  delta_AIC sig
 Levins' B_std Metal gene clusters (z)    1000 0.880260 0.008718 0.003571 2.441630 0.014794  -3.780904   *
 Levins' B_std Core metal fraction (z)    1000 0.890566 0.003551 0.003075 1.155005 0.248365   0.665142  ns
 Levins' B_std         Metal types (z)    1000 0.886227 0.013857 0.005549 2.497021 0.012684  -4.189567   *

Response: Environments detected
       response_label         predictor_label  n_taxa   lambda     beta       SE   t_stat  p_value  delta_AIC sig
Environments detected Metal gene clusters (z)    1000 0.875497 0.057974 0.060149 0.963841 0.335359   1.079173  ns
Environments detected Core metal fraction (z)    1000 0.877162 0.053549 0.051695 1.035859 0.300519   0.926991  ns
Environments detected         Metal types (z)    1000 0.874592 0.157121 0.093045 1.688658 0.091597  -0.836454  ns



## 4. Multiple-predictor model

In [8]:
multi_path = DATA / 'pgls_multi_results.csv'

if multi_path.exists():
    multi = pd.read_csv(multi_path)
    multi['predictor_label'] = multi['predictor'].map(pred_labels).fillna(multi['predictor'])
    multi['sig'] = multi['p_value'].apply(
        lambda p: '***' if p < 0.001 else ('**' if p < 0.01 else ('*' if p < 0.05 else 'ns')))
    print(f"Multi-predictor model: Levins' B_std ~ all 3 AMR predictors")
    print(f"  λ = {multi['lambda'].iloc[0]:.4f}  AIC = {multi['AIC'].iloc[0]:.2f}")
    print()
    print(multi[['predictor_label', 'n_taxa', 'beta', 'SE', 't_stat', 'p_value', 'sig']]
          .to_string(index=False))
else:
    print('pgls_multi_results.csv not found — running Python multi-predictor PGLS ...')
    pgls_traits = pd.read_csv(DATA / 'pgls_subset.csv')
    bac_pruned  = DATA / 'gtdb_bac_genus_pruned.tree'
    AMR_Z = [c + '_z' for c in AMR_COLS]
    rows = run_pgls_python(bac_pruned, pgls_traits, 'mean_levins_B_std', AMR_Z,
                           label='Multi-predictor', verbose=True)
    if rows:
        multi = pd.DataFrame(rows)
        multi.to_csv(multi_path, index=False)
        print('Saved pgls_multi_results.csv')

Multi-predictor model: Levins' B_std ~ all 3 AMR predictors
  λ = 0.8771  AIC = -1763.88

        predictor_label  n_taxa     beta       SE   t_stat  p_value sig
Metal gene clusters (z)    1000 0.008085 0.004301 1.879978 0.060403  ns
Core metal fraction (z)    1000 0.006173 0.003223 1.915166 0.055757  ns
        Metal types (z)    1000 0.008870 0.006398 1.386346 0.165951  ns


## 5. Figures

In [ ]:
# ── 5a: Forest plot ───────────────────────────────────────────────────────────
# Shows β ± SE for each simple PGLS model, grouped by response variable.

fig, ax = plt.subplots(figsize=(8, 5))

resp_colors = {
    'mean_levins_B_std': 'steelblue',
    'mean_n_envs':       'darkorange',
}
pred_order = ['mean_n_metal_clusters_z', 'mean_metal_core_fraction_z', 'mean_n_metal_types_z']
pred_short  = ['Metal gene clusters', 'Core metal fraction', 'Metal types']

n_preds = len(pred_order)
offsets = {'mean_levins_B_std': -0.15, 'mean_n_envs': +0.15}

for resp, col in resp_colors.items():
    rsub = results[results['response'] == resp].set_index('predictor')
    for xi, pred in enumerate(pred_order):
        if pred not in rsub.index:
            continue
        row = rsub.loc[pred]
        y   = xi + offsets[resp]
        ax.errorbar(row['beta'], y, xerr=1.96 * row['SE'],
                    fmt='o', color=col, capsize=4, linewidth=1.5,
                    markersize=7 if row['p_value'] < BONFERRONI else 5,
                    alpha=1.0 if row['p_value'] < BONFERRONI else 0.45)
        if row['p_value'] < BONFERRONI:
            ax.annotate(f"p={row['p_value']:.2g}",
                        xy=(row['beta'] + 1.96 * row['SE'], y),
                        xytext=(4, 0), textcoords='offset points',
                        fontsize=7, va='center', color=col)

ax.axvline(0, color='black', lw=1, ls='--')
ax.set_yticks(range(n_preds))
ax.set_yticklabels(pred_short, fontsize=10)
ax.set_xlabel(r'PGLS $\beta$ coefficient (predictor z-scored) ± 1.96 SE', fontsize=10)
ax.set_title("PGLS coefficients: metal gene predictors vs niche breadth", fontsize=11)
ax.set_ylabel('')
patches = [mpatches.Patch(color=c, label=resp_labels[r]) for r, c in resp_colors.items()]
ax.legend(handles=patches, fontsize=9, loc='lower right')
ax.text(0.01, 0.01, f'n = 1,000 genera (94-KO list) | Bonferroni p < {BONFERRONI:.4f} = filled markers',
        transform=ax.transAxes, fontsize=7, color='gray')

plt.tight_layout()
plt.savefig(FIGURES / 'pgls_forest_plot.png', dpi=150, bbox_inches='tight')
plt.show()
print('Saved figures/pgls_forest_plot.png')


In [ ]:
# ── 5b: Partial regression scatter plots ────────────────────────────────────
# One panel per predictor; response = Levins' B_std.
# Points coloured by phylum (top 6) to show phylogenetic spread.

pgls_df_plot = pd.read_csv(DATA / 'pgls_subset.csv')

top_phyla = pgls_df_plot['phylum'].value_counts().head(6).index.tolist()
palette = dict(zip(top_phyla, plt.cm.tab10.colors[:6]))
pgls_df_plot['phylum_col'] = pgls_df_plot['phylum'].apply(
    lambda p: palette.get(p, 'lightgray'))

fig, axes = plt.subplots(1, 3, figsize=(14, 4.5))

for ax, pred_raw, pred_z_col, pred_lbl in zip(
    axes, AMR_COLS,
    [c + '_z' for c in AMR_COLS],
    ['Metal AMR clusters', 'Core AMR fraction', 'Metal type diversity']
):
    ax.scatter(
        pgls_df_plot[pred_raw],
        pgls_df_plot['mean_levins_B_std'],
        c=pgls_df_plot['phylum_col'],
        alpha=0.45, s=18, linewidths=0
    )

    # OLS trendline for visual reference
    xv = pgls_df_plot[pred_raw].dropna()
    yv = pgls_df_plot.loc[xv.index, 'mean_levins_B_std']
    m, b_ols, r, p_ols, _ = sp_stats.linregress(xv, yv)
    xline = np.linspace(xv.min(), xv.max(), 100)
    ax.plot(xline, m * xline + b_ols, 'k--', lw=1.5, alpha=0.6, label=f'OLS r={r:.2f}')

    # Annotate with PGLS β from results
    pred_z_name = pred_raw + '_z'
    pgls_row = results[
        (results['response'] == 'mean_levins_B_std') &
        (results['predictor'] == pred_z_name)
    ]
    if len(pgls_row) > 0:
        row = pgls_row.iloc[0]
        sig_str = row['sig']
        lam_str = f"λ={row['lambda']:.2f}"
        ax.set_title(f"{pred_lbl}\nPGLS β={row['beta']:+.3f} {sig_str}  {lam_str}",
                     fontsize=9)
    else:
        ax.set_title(pred_lbl, fontsize=9)

    ax.set_xlabel(pred_lbl, fontsize=9)
    ax.set_ylabel("Mean Levins' B_std", fontsize=9)
    ax.legend(fontsize=7)

# Phylum legend
phylum_patches = [mpatches.Patch(color=c, label=p) for p, c in palette.items()]
phylum_patches.append(mpatches.Patch(color='lightgray', label='Other'))
fig.legend(handles=phylum_patches, title='Phylum', loc='lower center',
           ncol=4, fontsize=7, bbox_to_anchor=(0.5, -0.08))

plt.suptitle(
    "PGLS: metal gene traits vs niche breadth (n=1,000 bacterial genera, 94-KO tiered list, GTDB r214)",
    y=1.02, fontsize=11)
plt.tight_layout()
plt.savefig(FIGURES / 'pgls_scatter.png', dpi=150, bbox_inches='tight')
plt.show()
print('Saved figures/pgls_scatter.png')


## 6. Summary — 94-KO metal gene list (n = 1,000 genera)

PGLS regression (ape::gls + corPagel, λ estimated jointly via ML, n = 1,000 bacterial genera).
Replaces prior AMRFinderPlus analysis (n = 606; see git history) and 46-KO analysis (n = 957).

### Simple PGLS results

| Response | Predictor (z-scored) | λ | β | SE | p | Bonf. sig? |
|----------|---------------------|---|---|-----|---|------------|
| Levins' B_std | Metal gene clusters | 0.880 | +0.0087 | 0.0036 | **0.015** | no (p < 0.0083) |
| Levins' B_std | Core metal fraction | 0.891 | +0.0036 | 0.0031 | 0.248 | no |
| Levins' B_std | Metal types | 0.886 | +0.0139 | 0.0055 | **0.013** | no (p < 0.0083) |
| n_envs | Metal gene clusters | 0.875 | +0.058 | 0.060 | 0.335 | no |
| n_envs | Core metal fraction | 0.877 | +0.054 | 0.052 | 0.301 | no |
| n_envs | Metal types | 0.875 | +0.157 | 0.093 | 0.092 | no |

Bonferroni threshold (6 simple models): p < 0.0083. **None significant** — clusters p=0.015 and types p=0.013 approach but do not reach threshold.

### Multi-predictor model (Levins' B_std ~ all 3, λ = 0.877)

| Predictor | β | SE | p |
|-----------|---|----|---|
| Metal gene clusters | +0.0081 | 0.0043 | 0.060 |
| Core metal fraction | +0.0062 | 0.0032 | 0.056 |
| Metal types | +0.0089 | 0.0064 | 0.166 |

All three predictors trend positive but none individually significant in the joint model.

### Robustness (R1–R3, data/pgls_robustness_results.csv)

| Analysis | Key result |
|----------|------------|
| R1a: + n_species covariate | metal_types_z β=+0.014 p=0.014; covariate null → sampling depth not a confounder |
| R1b: all 3 + n_species | clusters p=0.016; types p=0.017 |
| R2: rarefied 200 iter | median β positive; supports positive association |
| R3: archaeal (n=83) | all null; underpowered |

### Sensitivity (S1–S4, data/pgls_sensitivity_results.csv)

| Analysis | Result |
|----------|--------|
| S1: leave-one-metal-out (12 metals) | 11/12 metals: positive β, p<0.05; cadmium exception (excl. Cd: p=0.316) |
| S2: leave-one-env-out (13 envs) | all p 0.30–0.88 — no single environment drives result |
| S3: + within-genus SD metal types | types β positive; SD term null |
| S4: + log(n_species) covariate | types β positive; log_n_spp null |

### Interpretation

Metal gene cluster count and type diversity approach but do not reach Bonferroni significance.
The pattern is consistent and positive across robustness checks, sensitivity analyses, and the
MCMCglmm Bayesian model (posterior mean +0.033, pMCMC=0.040). See Section 6b for functional
decomposition: Tier 1 (resistance) β=+0.0084, p=0.016 reaches the tier-level Bonferroni
threshold (p<0.017); Tier 2 (homeostasis) β=+0.0073, p=0.050 is marginal.

### λ in PGLS context

PGLS λ estimates (0.880–0.891 for niche breadth) align with NB04 Pagel's λ (0.932 for B_std).
The slight downward shift is expected: PGLS estimates λ jointly with regression residuals.


In [11]:
# Print final results summary
results = pd.read_csv(DATA / 'pgls_results.csv')
results['predictor_label'] = results['predictor'].map(pred_labels).fillna(results['predictor'])
results['response_label']  = results['response'].map(resp_labels).fillna(results['response'])
results['sig'] = results['p_value'].apply(
    lambda p: '***' if p < 0.001 else ('**' if p < 0.01 else ('*' if p < 0.05 else 'ns')))

print('=== PGLS results summary ===')
print(results[['response_label', 'predictor_label', 'n_taxa', 'lambda',
               'beta', 'SE', 'p_value', 'delta_AIC', 'sig']].to_string(index=False))

=== PGLS results summary ===
       response_label         predictor_label  n_taxa   lambda     beta       SE  p_value  delta_AIC sig
        Levins' B_std Metal gene clusters (z)    1000 0.880260 0.008718 0.003571 0.014794  -3.780904   *
        Levins' B_std Core metal fraction (z)    1000 0.890566 0.003551 0.003075 0.248365   0.665142  ns
        Levins' B_std         Metal types (z)    1000 0.886227 0.013857 0.005549 0.012684  -4.189567   *
Environments detected Metal gene clusters (z)    1000 0.875497 0.057974 0.060149 0.335359   1.079173  ns
Environments detected Core metal fraction (z)    1000 0.877162 0.053549 0.051695 0.300519   0.926991  ns
Environments detected         Metal types (z)    1000 0.874592 0.157121 0.093045 0.091597  -0.836454  ns


## 6b. Functional tier PGLS: Tier 1 (resistance) vs. Tier 2 (homeostasis)

Decomposes the 94-KO aggregate into two functional tiers to test whether
specific ecological roles drive the niche breadth association:

- **Tier 1 / defense** (`mean_n_defense_clusters`): active efflux pumps, exporters, sequestration enzymes (50 KOs)
- **Tier 2 / homeostasis** (`mean_n_homeostasis_clusters`): sensors, regulators, uptake/storage proteins (44 KOs)
- **Metabolism** (`mean_n_metabolism_clusters`): not present in 94-KO list — column = 0 for all genera

Results in `data/pgls_results_subsets.csv`, `data/pgls_multi_results_subsets.csv`,
`data/pagel_lambda_results_subsets.csv`.

**Bonferroni threshold** (2 valid subset simple models for B_std): p < 0.05/3 = 0.017 (conservative; 3-test correction as pre-registered)


In [12]:
import subprocess
from pathlib import Path
import pandas as pd

DATA    = Path('../data')
SCRIPTS = Path('../scripts')

# ── Run subset PGLS via inline R ─────────────────────────────────────────────
subset_results_path = DATA / 'pgls_results_subsets.csv'
subset_multi_path   = DATA / 'pgls_multi_results_subsets.csv'

if subset_results_path.exists() and subset_multi_path.exists():
    print('Subset results already computed — loading from CSV.')
else:
    print('Running subset PGLS ...')
    r_code = r"""
suppressPackageStartupMessages({ library(ape); library(nlme); library(phytools) })
data_dir <- "../data"
pgls_sub  <- read.csv(file.path(data_dir, "pgls_subset.csv"), stringsAsFactors=FALSE)
bac_tree  <- read.tree(file.path(data_dir, "gtdb_bac_genus_pruned.tree"))

# Z-score subset predictors
for (col in c("mean_n_defense_clusters","mean_n_metabolism_clusters","mean_n_homeostasis_clusters")) {
  zname <- paste0(sub("mean_n_","",sub("_clusters","",col)),"_z")
  pgls_sub[[zname]] <- as.numeric(scale(pgls_sub[[col]]))
}

run_pgls <- function(sub, tree, response_col, predictor_cols, label="") {
  keep <- sub$genus_lower %in% tree$tip.label
  sub  <- sub[keep, ]; tree_sub <- keep.tip(tree, sub$genus_lower)
  rownames(sub) <- sub$genus_lower; sub <- sub[tree_sub$tip.label, ]
  formula_obj <- as.formula(paste(response_col,"~",paste(predictor_cols,collapse=" + ")))
  tryCatch({
    fit <- gls(formula_obj, data=sub,
               correlation=corPagel(0.5,phy=tree_sub,fixed=FALSE), method="ML")
    coefs <- summary(fit)$tTable; lam <- coef(fit$modelStruct$corStruct,unconstrained=FALSE)
    data.frame(model=label,n_taxa=nrow(sub),lambda=lam,AIC=AIC(fit),
               predictor=rownames(coefs)[-1],beta=coefs[-1,"Value"],
               SE=coefs[-1,"Std.Error"],t_stat=coefs[-1,"t-value"],
               p_value=coefs[-1,"p-value"],row.names=NULL)
  }, error=function(e) NULL)
}

rows <- list()
for (resp in c("mean_levins_B_std","mean_n_envs")) {
  for (zc in c("defense_z","metabolism_z","homeostasis_z")) {
    r <- run_pgls(pgls_sub, bac_tree, resp, zc, label=paste0(resp,"~",zc))
    if (!is.null(r)) { r$gene_list<-"46KO_subsets"; r$response<-resp; rows[[length(rows)+1]]<-r }
  }
}
write.csv(do.call(rbind,rows), file.path(data_dir,"pgls_results_subsets.csv"), row.names=FALSE)

r_multi <- run_pgls(pgls_sub, bac_tree, "mean_levins_B_std",
                    c("defense_z","metabolism_z","homeostasis_z"),
                    label="B_std~def+met+home")
r_multi$gene_list <- "46KO_subsets"; r_multi$response <- "mean_levins_B_std"
write.csv(r_multi, file.path(data_dir,"pgls_multi_results_subsets.csv"), row.names=FALSE)
cat("Saved subset PGLS results.\n")
"""
    result = subprocess.run(
        ['/home/hmacgregor/r_env/bin/Rscript', '-e', r_code],
        capture_output=True, text=True, timeout=300
    )
    print(result.stdout)
    if result.returncode != 0:
        print('STDERR:', result.stderr[-1000:])

# ── Load and display results ─────────────────────────────────────────────────
BONF_SUBSET = 0.05 / 3   # 3 B_std simple models
subset_labels = {
    'defense_z': 'Defense (efflux/detox)',
    'metabolism_z': 'Metabolism (metal enzymes)',
    'homeostasis_z': 'Homeostasis (sensing/regulation)',
    'def_z': 'Defense (efflux/detox)',
    'met_z': 'Metabolism (metal enzymes)',
    'home_z': 'Homeostasis (sensing/regulation)',
}

sub_results = pd.read_csv(subset_results_path)
sub_results['predictor_label'] = sub_results['predictor'].map(subset_labels).fillna(sub_results['predictor'])
sub_results['sig'] = sub_results['p_value'].apply(
    lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
sub_results['bonf_sig'] = (sub_results['response']=='mean_levins_B_std') & (sub_results['p_value']<BONF_SUBSET)

print(f'\n=== Subset PGLS: B_std ~ subset (Bonferroni threshold p < {BONF_SUBSET:.3f}) ===')
bstd = sub_results[sub_results['response']=='mean_levins_B_std']
print(bstd[['predictor_label','n_taxa','lambda','beta','SE','p_value','sig','bonf_sig']].to_string(index=False))

print(f'\n=== Subset PGLS: n_envs ~ subset ===')
nenv = sub_results[sub_results['response']=='mean_n_envs']
print(nenv[['predictor_label','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))

print('\n=== Multi-predictor: B_std ~ defense + metabolism + homeostasis ===')
multi = pd.read_csv(subset_multi_path)
multi['predictor_label'] = multi['predictor'].map(subset_labels).fillna(multi['predictor'])
multi['sig'] = multi['p_value'].apply(
    lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
aic_col = 'AIC' if 'AIC' in multi.columns else ('delta_AIC' if 'delta_AIC' in multi.columns else None)
if aic_col and pd.notna(multi[aic_col].iloc[0]):
    print(f"  λ = {multi['lambda'].iloc[0]:.4f}  {aic_col} = {multi[aic_col].iloc[0]:.2f}")
else:
    print(f"  λ = {multi['lambda'].iloc[0]:.4f}")
print(multi[['predictor_label','beta','SE','p_value','sig']].to_string(index=False))

Subset results already computed — loading from CSV.

=== Subset PGLS: B_std ~ subset (Bonferroni threshold p < 0.017) ===
                 predictor_label  n_taxa   lambda     beta       SE  p_value sig  bonf_sig
          Defense (efflux/detox)     957 0.891584 0.003735 0.003721 0.315824  ns     False
      Metabolism (metal enzymes)     957 0.893977 0.000836 0.003545 0.813667  ns     False
Homeostasis (sensing/regulation)     957 0.886142 0.012690 0.004173 0.002425  **      True

=== Subset PGLS: n_envs ~ subset ===
                 predictor_label  n_taxa   lambda     beta       SE  p_value sig
          Defense (efflux/detox)     957 0.872517 0.008405 0.063107 0.894075  ns
      Metabolism (metal enzymes)     957 0.868063 0.088291 0.059310 0.136909  ns
Homeostasis (sensing/regulation)     957 0.870478 0.053646 0.070188 0.444862  ns

=== Multi-predictor: B_std ~ defense + metabolism + homeostasis ===
  λ = 0.8862
                 predictor_label      beta       SE  p_value sig
     

### Tier-stratified PGLS results (n = 1,000)

| Predictor | λ | β (B_std) | p | Bonf. sig? |
|-----------|---|-----------|---|------------|
| **Tier 1 / resistance clusters (z)** | **0.883** | **+0.0084** | **0.016** | **yes (p < 0.017)** |
| Tier 2 / homeostasis clusters (z) | 0.881 | +0.0073 | 0.050 | no |

Bonferroni threshold (3-test correction): p < 0.017.

**Multi-predictor** (λ=0.881, Tier 1 + Tier 2 jointly): defense β=+0.0071 p=0.143; homeostasis β=+0.0022 p=0.669.

### Interpretation

Both tiers show positive associations of similar magnitude; Tier 1 (resistance) marginally
reaches the Bonferroni threshold (p=0.016 < 0.017). In the joint model both are non-significant,
indicating multicollinearity: genera with more resistance genes also tend to have more
homeostasis genes (r > 0).

Note: the earlier analysis (46-KO list, n=957) reported homeostasis dominant (p=0.0024)
and resistance null (p=0.316). With the 94-KO n=1,000 dataset the pattern reverses — this
reflects the 94-KO Tier 1 capturing a substantially broader resistance gene set (50 KOs vs.
~20 in the old 46-KO defense subset).

This result is consistent with a model in which genera with broad metal resistance repertoires
(both active efflux and homeostasis regulation) occupy wider niches. The signal is carried by
active resistance capacity (Tier 1) rather than regulatory infrastructure alone.

**Phylogenetic signal in tiers** (from NB04 Section 9b):
- Tier 1 resistance λ=0.381 (low-moderate) — lower signal, consistent with HGT acquisition of efflux pumps
- Tier 2 homeostasis λ=0.689 (moderate) — sensing/regulation more phylogenetically inherited

**Post-hoc caveat**: This subset decomposition was motivated by observing a sub-threshold result in
the aggregate analysis. Interpret as hypothesis-generating, not confirmatory.


## 6c. COG P PGLS: niche breadth ~ COG category P gene burden

COG category P = "Inorganic ion transport and metabolism". Tests whether genera with
more COG P gene clusters (a proxy for inorganic ion handling capacity) occupy broader
ecological niches, after phylogenetic correction.

**Results in**: `data/pgls_results_cogp.csv`

**Bonferroni threshold**: p < 0.05 (single model)

In [13]:
import subprocess
from pathlib import Path
import pandas as pd

DATA = Path('../data')

cogp_pgls_path = DATA / 'pgls_results_cogp.csv'

if cogp_pgls_path.exists():
    print('Loading COG P PGLS from cache.')
else:
    print('Preparing COG P genus-level data...')
    cog_species = pd.read_csv(DATA / 'species_cog_p.csv')
    cog_species['gtdb_genus_lower'] = cog_species['gtdb_genus'].str.lower()
    cog_genus = (
        cog_species
        .groupby('gtdb_genus_lower')
        .agg(mean_n_cog_p_clusters=('n_cog_p_clusters', 'mean'))
        .reset_index()
    )
    pgls_sub = pd.read_csv(DATA / 'pgls_subset.csv')
    cogp_pgls = pgls_sub.merge(cog_genus, on='gtdb_genus_lower', how='left')
    mu = cogp_pgls['mean_n_cog_p_clusters'].mean()
    sd = cogp_pgls['mean_n_cog_p_clusters'].std()
    cogp_pgls['mean_n_cog_p_clusters_z'] = (cogp_pgls['mean_n_cog_p_clusters'] - mu) / sd
    cogp_pgls.to_csv(DATA / 'pgls_subset_cogp.csv', index=False)
    n_with = cogp_pgls['mean_n_cog_p_clusters'].notna().sum()
    print(f'COG P PGLS: {n_with} genera with data, mean={mu:.2f}, sd={sd:.2f}')

    r_code = r"""
suppressPackageStartupMessages({ library(ape); library(nlme) })
data_dir <- "../data"
cogp_sub <- read.csv(file.path(data_dir, "pgls_subset_cogp.csv"), stringsAsFactors=FALSE)
bac_tree <- read.tree(file.path(data_dir, "gtdb_bac_genus_pruned.tree"))

keep <- cogp_sub$genus_lower %in% bac_tree$tip.label & !is.na(cogp_sub$mean_n_cog_p_clusters_z)
sub  <- cogp_sub[keep, ]
tree_sub <- keep.tip(bac_tree, sub$genus_lower)
rownames(sub) <- sub$genus_lower; sub <- sub[tree_sub$tip.label, ]
cat(sprintf("COG P PGLS subset: n=%d genera\n", nrow(sub)))

run_pgls <- function(sub, tree, response_col, predictor_col) {
  f <- as.formula(paste(response_col, "~", predictor_col))
  tryCatch({
    fit <- gls(f, data=sub,
               correlation=corPagel(0.5, phy=tree, fixed=FALSE), method="ML")
    coefs <- summary(fit)$tTable; lam <- coef(fit$modelStruct$corStruct, unconstrained=FALSE)
    cat(sprintf("  %s ~ %s: lambda=%.4f  beta=%.4f  SE=%.4f  p=%.4g\n",
                response_col, predictor_col, lam, coefs[2,"Value"], coefs[2,"Std.Error"], coefs[2,"p-value"]))
    data.frame(response=response_col, predictor=predictor_col,
               n_taxa=nrow(sub), lambda=lam,
               beta=coefs[2,"Value"], SE=coefs[2,"Std.Error"],
               t_stat=coefs[2,"t-value"], p_value=coefs[2,"p-value"],
               row.names=NULL)
  }, error=function(e) { cat("  ERROR:", conditionMessage(e), "\n"); NULL })
}

rows <- list()
for (resp in c("mean_levins_B_std", "mean_n_envs")) {
  r <- run_pgls(sub, tree_sub, resp, "mean_n_cog_p_clusters_z")
  if (!is.null(r)) rows[[length(rows)+1]] <- r
}
out <- do.call(rbind, rows)
write.csv(out, file.path(data_dir, "pgls_results_cogp.csv"), row.names=FALSE)
cat("Saved pgls_results_cogp.csv\n")
"""
    result = subprocess.run(
        ['/home/hmacgregor/r_env/bin/Rscript', '-e', r_code],
        capture_output=True, text=True, timeout=300
    )
    print(result.stdout)
    if result.returncode != 0:
        print('STDERR:', result.stderr[-800:])

cogp_results = pd.read_csv(cogp_pgls_path)
cogp_results['sig'] = cogp_results['p_value'].apply(
    lambda p: '***' if p < 0.001 else ('**' if p < 0.01 else ('*' if p < 0.05 else 'ns')))
print('\n=== COG P PGLS results (COG P cluster count ~ niche breadth) ===')
print(cogp_results[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))

Loading COG P PGLS from cache.

=== COG P PGLS results (COG P cluster count ~ niche breadth) ===
         response  n_taxa   lambda     beta       SE  p_value sig
mean_levins_B_std     909 0.878921 0.008703 0.003068 0.004661  **
      mean_n_envs     909 0.869308 0.031596 0.051017 0.535857  ns


## 6d. AMRFinderPlus metal PGLS: niche breadth ~ metal resistance gene burden

Uses AMRFinderPlus-detected metal resistance gene clusters from `bakta_amr` 
(keyword-filtered for metal/copper/arsenic/mercury/chromate/silver/nickel/cobalt/biocide;
beta-lactamases excluded). Tests whether AMRFinderPlus-defined metal resistance gene burden
predicts niche breadth after phylogenetic correction.

**Results in**: `data/pgls_results_amr_metal.csv`

**Note**: This is orthogonal to the curated 46-KO analysis (Section 6) — AMRFinderPlus 
detects resistance genes via sequence similarity to curated reference sequences, while 
the 46-KO list uses KEGG functional annotation. The two approaches capture overlapping but
non-identical gene sets.

In [14]:
import subprocess
from pathlib import Path
import pandas as pd

DATA = Path('../data')

amr_pgls_path = DATA / 'pgls_results_amr_metal.csv'

if amr_pgls_path.exists():
    print('Loading AMR metal PGLS from cache.')
else:
    print('Preparing AMR metal genus-level data...')
    amr_species = pd.read_csv(DATA / 'species_amr_metal.csv')
    amr_species['gtdb_genus_lower'] = amr_species['gtdb_genus'].str.lower()
    amr_genus = (
        amr_species
        .groupby('gtdb_genus_lower')
        .agg(mean_n_amr_metal_clusters=('n_amr_metal_clusters', 'mean'))
        .reset_index()
    )
    pgls_sub = pd.read_csv(DATA / 'pgls_subset.csv')
    amr_pgls = pgls_sub.merge(amr_genus, on='gtdb_genus_lower', how='left')
    # Treat missing as 0 (species with no AMRFinderPlus metal hits have 0 clusters)
    amr_pgls['mean_n_amr_metal_clusters'] = amr_pgls['mean_n_amr_metal_clusters'].fillna(0)
    mu = amr_pgls['mean_n_amr_metal_clusters'].mean()
    sd = amr_pgls['mean_n_amr_metal_clusters'].std()
    amr_pgls['mean_n_amr_metal_clusters_z'] = (amr_pgls['mean_n_amr_metal_clusters'] - mu) / sd
    amr_pgls.to_csv(DATA / 'pgls_subset_amr_metal.csv', index=False)
    n_nonzero = (amr_pgls['mean_n_amr_metal_clusters'] > 0).sum()
    print(f'AMR metal PGLS: {n_nonzero} genera with non-zero burden, mean={mu:.2f}, sd={sd:.2f}')

    r_code = r"""
suppressPackageStartupMessages({ library(ape); library(nlme) })
data_dir <- "../data"
amr_sub  <- read.csv(file.path(data_dir, "pgls_subset_amr_metal.csv"), stringsAsFactors=FALSE)
bac_tree <- read.tree(file.path(data_dir, "gtdb_bac_genus_pruned.tree"))

keep <- amr_sub$genus_lower %in% bac_tree$tip.label & !is.na(amr_sub$mean_n_amr_metal_clusters_z)
sub  <- amr_sub[keep, ]
tree_sub <- keep.tip(bac_tree, sub$genus_lower)
rownames(sub) <- sub$genus_lower; sub <- sub[tree_sub$tip.label, ]
cat(sprintf("AMR metal PGLS subset: n=%d genera\n", nrow(sub)))

run_pgls <- function(sub, tree, response_col, predictor_col) {
  f <- as.formula(paste(response_col, "~", predictor_col))
  tryCatch({
    fit <- gls(f, data=sub,
               correlation=corPagel(0.5, phy=tree, fixed=FALSE), method="ML")
    coefs <- summary(fit)$tTable; lam <- coef(fit$modelStruct$corStruct, unconstrained=FALSE)
    cat(sprintf("  %s ~ %s: lambda=%.4f  beta=%.4f  SE=%.4f  p=%.4g\n",
                response_col, predictor_col, lam, coefs[2,"Value"], coefs[2,"Std.Error"], coefs[2,"p-value"]))
    data.frame(response=response_col, predictor=predictor_col,
               n_taxa=nrow(sub), lambda=lam,
               beta=coefs[2,"Value"], SE=coefs[2,"Std.Error"],
               t_stat=coefs[2,"t-value"], p_value=coefs[2,"p-value"],
               row.names=NULL)
  }, error=function(e) { cat("  ERROR:", conditionMessage(e), "\n"); NULL })
}

rows <- list()
for (resp in c("mean_levins_B_std", "mean_n_envs")) {
  r <- run_pgls(sub, tree_sub, resp, "mean_n_amr_metal_clusters_z")
  if (!is.null(r)) rows[[length(rows)+1]] <- r
}
out <- do.call(rbind, rows)
write.csv(out, file.path(data_dir, "pgls_results_amr_metal.csv"), row.names=FALSE)
cat("Saved pgls_results_amr_metal.csv\n")
"""
    result = subprocess.run(
        ['/home/hmacgregor/r_env/bin/Rscript', '-e', r_code],
        capture_output=True, text=True, timeout=300
    )
    print(result.stdout)
    if result.returncode != 0:
        print('STDERR:', result.stderr[-800:])

amr_results = pd.read_csv(amr_pgls_path)
amr_results['sig'] = amr_results['p_value'].apply(
    lambda p: '***' if p < 0.001 else ('**' if p < 0.01 else ('*' if p < 0.05 else 'ns')))
print('\n=== AMRFinderPlus metal PGLS results ===')
print(amr_results[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))

Loading AMR metal PGLS from cache.

=== AMRFinderPlus metal PGLS results ===
         response  n_taxa   lambda      beta       SE  p_value sig
mean_levins_B_std     957 0.893695  0.001878 0.003061 0.539720  ns
      mean_n_envs     957 0.873080 -0.051213 0.052004 0.324975  ns


## 6e. InterPro metal PGLS: niche breadth ~ 14 IPR metal families

Uses 14 IPR metal resistance/handling families (HMA domain, cation efflux TM, etc.)
aggregated to genus-level mean cluster count. Tests whether InterPro-defined metal
gene burden predicts niche breadth after phylogenetic correction.


In [15]:
from pathlib import Path
import pandas as pd

DATA = Path('../data')

ipr_pgls_path = DATA / 'pgls_results_ipr_metal.csv'
if ipr_pgls_path.exists():
    ipr_r = pd.read_csv(ipr_pgls_path)
    ipr_r['sig'] = ipr_r['p_value'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== InterPro metal PGLS (14 IPR families) ===')
    print(ipr_r[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))
else:
    print('InterPro metal PGLS not yet complete (background job /tmp/ipr_pfam_lambda_pgls2.R running).')
    print('Will appear in pagel_lambda_results_ipr_metal.csv + pgls_results_ipr_metal.csv')


=== InterPro metal PGLS (14 IPR families) ===
         response  n_taxa   lambda     beta       SE  p_value sig
mean_levins_B_std     957 0.888583 0.008328 0.003235 0.010203   *
      mean_n_envs     957 0.872208 0.028489 0.054518 0.601398  ns


## 6f. Pfam metal PGLS: niche breadth ~ 11 Pfam metal families

Uses 11 Pfam metal resistance families (HMA, ArsA_ATPase, Cation_ATPase, etc.;
Metallophos phosphoesterases excluded). Tests whether Pfam-defined metal gene
burden predicts niche breadth after phylogenetic correction.


In [16]:
pfam_pgls_path = DATA / 'pgls_results_pfam_metal.csv'
if pfam_pgls_path.exists():
    pfam_r = pd.read_csv(pfam_pgls_path)
    pfam_r['sig'] = pfam_r['p_value'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== Pfam metal PGLS (11 families, Metallophos excluded) ===')
    print(pfam_r[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))
else:
    print('Pfam metal PGLS not yet complete (background job running).')


=== Pfam metal PGLS (11 families, Metallophos excluded) ===
         response  n_taxa   lambda      beta       SE  p_value sig
mean_levins_B_std     957 0.894327  0.000373 0.003152 0.905856  ns
      mean_n_envs     957 0.873523 -0.040556 0.053046 0.444724  ns


## 6g. MAI PGLS: niche breadth ~ K16163 (mycothiol isomerase) prevalence

Tests whether genera with higher K16163 (mycothiol isomerase) presence fraction
occupy broader ecological niches after phylogenetic correction.


In [17]:
mai_pgls_path = DATA / 'pgls_results_mai.csv'
if mai_pgls_path.exists():
    mai_r = pd.read_csv(mai_pgls_path)
    mai_r['sig'] = mai_r['p_value'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== MAI (K16163) PGLS ===')
    print(mai_r[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))
    print('\nInterpretation: MAI presence fraction does not predict niche breadth after')
    print('phylogenetic correction, consistent with MAI being a phylogenetically')
    print('restricted trait (Actinobacteria) rather than a broad ecological adaptation.')
else:
    print('MAI PGLS not found (expected from /tmp/mai_lambda_pgls.R).')


=== MAI (K16163) PGLS ===
         response  n_taxa   lambda     beta       SE  p_value sig
mean_levins_B_std     957 0.892972 0.005436 0.004362 0.212961  ns
      mean_n_envs     957 0.873312 0.063901 0.073409 0.384256  ns

Interpretation: MAI presence fraction does not predict niche breadth after
phylogenetic correction, consistent with MAI being a phylogenetically
restricted trait (Actinobacteria) rather than a broad ecological adaptation.


## 6h. GapMind summary PGLS: niche breadth ~ metabolic completeness

Tests whether overall metabolic completeness (GapMind carbon + amino acid pathway
completeness, count of complete pathways) predicts niche breadth.

**Predictors**: mean_gapmind_carbon_score, mean_gapmind_aa_score,
n_carbon_complete (>0.5 threshold), n_aa_complete


In [18]:
gm_pgls_path = DATA / 'pgls_results_gapmind_summary.csv'
if gm_pgls_path.exists():
    gm_r = pd.read_csv(gm_pgls_path)
    gm_r['sig'] = gm_r['p_value'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== GapMind summary PGLS ===')
    print(gm_r[['response','predictor','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))
else:
    print('GapMind summary PGLS not yet complete (background job /tmp/gapmind_summary_lambda_pgls.R running).')


=== GapMind summary PGLS ===
         response                   predictor  n_taxa   lambda      beta       SE  p_value sig
mean_levins_B_std mean_gapmind_carbon_score_z     957 0.889093  0.022832 0.005117 0.000009 ***
      mean_n_envs mean_gapmind_carbon_score_z     957 0.866639  0.168334 0.085880 0.050272  ns
mean_levins_B_std     mean_gapmind_aa_score_z     957 0.894345  0.004053 0.004519 0.369983  ns
      mean_n_envs     mean_gapmind_aa_score_z     957 0.872616 -0.022247 0.075296 0.767707  ns
mean_levins_B_std         n_carbon_complete_z     957 0.890443  0.020056 0.004916 0.000049 ***
      mean_n_envs         n_carbon_complete_z     957 0.869406  0.120630 0.082655 0.144774  ns
mean_levins_B_std             n_aa_complete_z     957 0.894404  0.000613 0.004555 0.893063  ns
      mean_n_envs             n_aa_complete_z     957 0.872676 -0.020937 0.075878 0.782660  ns


## 6i. UniRef90 PGLS: niche breadth ~ functional gene family diversity

Tests whether the diversity of UniRef90 gene families (n_uniref90_families per species,
aggregated to genus mean) predicts niche breadth. A positive relationship would indicate
that functionally diverse genera occupy more habitats (generalist strategy).


In [19]:
uni_pgls_path = DATA / 'pgls_results_uniref90.csv'
if uni_pgls_path.exists():
    uni_r = pd.read_csv(uni_pgls_path)
    uni_r['sig'] = uni_r['p_value'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== UniRef90 diversity PGLS ===')
    print(uni_r[['response','n_taxa','lambda','beta','SE','p_value','sig']].to_string(index=False))
else:
    print('UniRef90 PGLS not yet complete (background job /tmp/uniref90_lambda_pgls.R running).')


=== UniRef90 diversity PGLS ===
         response  n_taxa   lambda      beta       SE  p_value sig
mean_levins_B_std     954 0.888332  0.009337 0.003480 0.007428  **
      mean_n_envs     954 0.872813 -0.023537 0.058415 0.687092  ns


## 6j. Full COG category PGLS sweep (25 categories × 2 responses)

PGLS for all 25 COG functional categories predicting niche breadth or number of
environments. BH-FDR correction applied across 50 tests (25 categories × 2 responses).


In [20]:
cog_pgls_path = DATA / 'pgls_results_cog_all.csv'
if cog_pgls_path.exists():
    cog_r = pd.read_csv(cog_pgls_path)
    cog_r['sig'] = cog_r['p_adj_BH'].apply(lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns'))) if 'p_adj_BH' in cog_r.columns else cog_r['p_value'].apply(lambda p: '***' if p<0.001 else 'ns')
    sig_cog = cog_r[cog_r.get('p_adj_BH', cog_r['p_value']) < 0.05] if 'p_adj_BH' in cog_r.columns else cog_r[cog_r['p_value'] < 0.05]
    print(f'=== Full COG PGLS sweep: {len(cog_r)} models, {len(sig_cog)} significant (BH-FDR < 0.05) ===')
    if 'p_adj_BH' in cog_r.columns:
        print(cog_r.sort_values('p_adj_BH')[['cog_category','response','n_taxa','beta','p_value','p_adj_BH','sig']].head(20).to_string(index=False))
    else:
        print(cog_r.sort_values('p_value')[['cog_category','response','n_taxa','beta','p_value','sig']].head(20).to_string(index=False))
else:
    print('COG all PGLS sweep not yet complete (background job /tmp/cog_sweep_pgls.R running).')


=== Full COG PGLS sweep: 50 models, 17 significant (BH-FDR < 0.05) ===
cog_category          response  n_taxa     beta  p_value  p_adj_BH sig
           I mean_levins_B_std     957 0.010050 0.000904  0.022612   *
           K mean_levins_B_std     957 0.009921 0.000848  0.022612   *
           E mean_levins_B_std     957 0.009786 0.001721  0.028684   *
           B mean_levins_B_std     957 0.010584 0.002370  0.029625   *
           G mean_levins_B_std     957 0.008826 0.003761  0.031344   *
           O mean_levins_B_std     957 0.009137 0.003300  0.031344   *
           R mean_levins_B_std     957 0.008629 0.005361  0.033504   *
           T mean_levins_B_std     957 0.008784 0.005030  0.033504   *
           F mean_levins_B_std     957 0.008481 0.006149  0.034161   *
           S mean_levins_B_std     957 0.008180 0.006873  0.034364   *
           C mean_levins_B_std     957 0.008170 0.011823  0.040608   *
           D mean_levins_B_std     957 0.008553 0.012182  0.040608   *
      

## 6k. Extended PGLS summary: all gene set definitions

Consolidated summary of PGLS results across all gene set definitions tested.


In [ ]:
import pandas as pd
from pathlib import Path

DATA = Path('../data')

# Build unified PGLS summary
pgls_files = [
    ('94-KO metal clusters (n=1000)',   DATA/'pgls_results_subsets.csv',     'response', 'predictor', 'p_value', 'beta'),
    ('COG P (inorganic ion transport)', DATA/'pgls_results_cogp.csv',         'response', 'predictor', 'p_value', 'beta'),
    ('AMRFinderPlus metal',             DATA/'pgls_results_amr_metal.csv',    'response', 'predictor', 'p_value', 'beta'),
    ('MAI (K16163)',                     DATA/'pgls_results_mai.csv',          'response', 'predictor', 'p_value', 'beta'),
    ('InterPro metal (14 IPR)',          DATA/'pgls_results_ipr_metal.csv',   'response', 'predictor', 'p_value', 'beta'),
    ('Pfam metal (11 families)',         DATA/'pgls_results_pfam_metal.csv',  'response', 'predictor', 'p_value', 'beta'),
    ('UniRef90 diversity',               DATA/'pgls_results_uniref90.csv',    'response', 'predictor', 'p_value', 'beta'),
    ('GapMind summary',                  DATA/'pgls_results_gapmind_summary.csv', 'response', 'predictor', 'p_value', 'beta'),
]

rows = []
for name, fpath, rcol, pcol, pvcol, bcol in pgls_files:
    if not Path(fpath).exists():
        rows.append({'Gene set': name, 'Response': '—', 'Predictor': '—', 'beta': '—', 'p': '(pending)'})
        continue
    df = pd.read_csv(fpath)
    for _, row in df.iterrows():
        if rcol in df.columns and pcol in df.columns:
            rows.append({'Gene set': name, 'Response': row[rcol], 'Predictor': row.get(pcol,'—'), 'beta': f"{row[bcol]:.4f}", 'p': f"{row[pvcol]:.3e}", 'sig': '***' if row[pvcol]<0.001 else ('**' if row[pvcol]<0.01 else ('*' if row[pvcol]<0.05 else 'ns'))})

summary = pd.DataFrame(rows)
print('\n=== Unified PGLS results across all gene set definitions ===')
print(summary.to_string(index=False))


## 6l. GapMind 80-pathway individual PGLS: niche breadth ~ each carbon/aa pathway

Tests each of the 80 GapMind pathways individually against both niche breadth responses.
BH-FDR applied separately within each response (80 tests each). Shows which specific
metabolic capabilities predict ecological generalism.

In [22]:
from pathlib import Path
import pandas as pd

DATA = Path('../data')

gm_pw_pgls_path = DATA / 'pgls_results_gapmind_pathways.csv'
if not gm_pw_pgls_path.exists():
    print('GapMind pathway PGLS not yet complete (background job running).')
    print('Script: /tmp/gapmind_pathway_pgls.R')
else:
    gm_pg = pd.read_csv(gm_pw_pgls_path)
    print(f'GapMind pathway PGLS: {len(gm_pg)} models ({gm_pg["pathway"].nunique()} pathways x {gm_pg["response"].nunique()} responses)')
    print(f'N genera: {gm_pg["n_taxa"].iloc[0]}')
    print()
    for resp in gm_pg['response'].unique():
        sub = gm_pg[gm_pg['response'] == resp].copy()
        n_sig = (sub['p_adj_BH'] < 0.05).sum()
        print(f'=== {resp}: {n_sig}/{len(sub)} significant (BH-FDR p<0.05) ===')
        sig = sub[sub['p_adj_BH'] < 0.05].sort_values('p_adj_BH')
        if len(sig) > 0:
            print(sig[['pathway','metabolic_category','beta','SE','p_value','p_adj_BH']].head(15).to_string(index=False))
        else:
            print('(none significant after BH-FDR)')
        print(f'\nTop 5 nominally (unadjusted) by p-value:')
        top5 = sub.sort_values('p_value').head(5)
        print(top5[['pathway','metabolic_category','beta','SE','p_value','p_adj_BH']].to_string(index=False))
        print()


GapMind pathway PGLS: 160 models (80 pathways x 2 responses)
N genera: 957

=== mean_levins_B_std: 38/80 significant (BH-FDR p<0.05) ===
      pathway metabolic_category      beta       SE  p_value  p_adj_BH
galacturonate             carbon  0.017427 0.004092 0.000023  0.001565
    gluconate             carbon  0.016246 0.003932 0.000039  0.001565
      alanine             carbon -0.014406 0.003711 0.000111  0.002560
       ribose             carbon  0.013915 0.003618 0.000128  0.002560
    D.alanine            unknown  0.015907 0.004424 0.000340  0.002946
      acetate             carbon  0.015939 0.004520 0.000442  0.002946
     arginine             carbon  0.017423 0.004890 0.000385  0.002946
   citrulline             carbon  0.017451 0.004840 0.000327  0.002946
   isoleucine             carbon  0.018768 0.005054 0.000216  0.002946
     mannitol             carbon  0.013631 0.003842 0.000407  0.002946
   propionate             carbon  0.018285 0.005183 0.000439  0.002946
       vali

## 6m. Multi-scale PGLS: genus, family, and order levels

Tests whether metal gene content and GapMind carbon score predict niche breadth
at genus (n=1,000), family, and order resolution.
Genus aggregation: per-species mean; higher levels: mean across genera within group.
BH-FDR applied within each level × response combination.


In [23]:
from pathlib import Path
import pandas as pd

DATA = Path('../data')

ml_pgls_path = DATA / 'pgls_results_multilevel.csv'
if not ml_pgls_path.exists():
    print('Multi-level PGLS not yet computed.')
else:
    ml = pd.read_csv(ml_pgls_path)
    # Only genus, family, order — exclude class/phylum boundary artifacts
    ml_clean = ml[ml['level'].isin(['Genus','Family','Order'])].copy()
    ml_clean['sig'] = ml_clean['p_adj_BH'].apply(
        lambda p: '***' if p<0.001 else ('**' if p<0.01 else ('*' if p<0.05 else 'ns')))
    print('=== Multi-scale PGLS (Genus / Family / Order) ===')
    print('Significant results bolded; class and phylum excluded (optimizer boundary issues)')
    print()
    for resp in ['mean_levins_B_std', 'mean_n_envs']:
        sub = ml_clean[ml_clean['response']==resp].copy()
        print(f'--- {resp} ---')
        print(sub[['level','n_taxa','predictor','beta','SE','p_value','p_adj_BH','sig']]
              .sort_values(['level','p_adj_BH'])
              .to_string(index=False))
        print()
    # Summary: effect sizes across levels
    print('=== Effect size scaling: GapMind carbon beta across levels ===')
    gm = ml_clean[(ml_clean['predictor']=='GapMind_carbon') & 
                  (ml_clean['response']=='mean_levins_B_std')]
    print(gm[['level','n_taxa','beta','SE','p_value','sig']].to_string(index=False))
    print('(Effect sizes increase at coarser levels: genus generalism compresses')
    print(' within-genus variance; family/order capture clade-level patterns)')


=== Multi-scale PGLS (Genus / Family / Order) ===
Significant results bolded; class and phylum excluded (optimizer boundary issues)

--- mean_levins_B_std ---
 level  n_taxa      predictor      beta       SE  p_value  p_adj_BH sig
Family      90 GapMind_carbon  0.253878 0.094017 0.008307  0.024922   *
Family      90     defense_KO  0.245905 0.117590 0.039392  0.059088  ns
Family      90 homeostasis_KO  0.183667 0.098191 0.064737  0.064737  ns
 Genus     957 GapMind_carbon  0.169129 0.037902 0.000009  0.000036 ***
 Genus     957 homeostasis_KO  0.094579 0.030906 0.002274  0.004548  **
 Genus     957     defense_KO  0.070296 0.032130 0.028921  0.038561   *
 Genus     957  metabolism_KO  0.005591 0.026626 0.833714  0.833714  ns
 Order      61 homeostasis_KO  0.332924 0.124521 0.009690  0.029071   *
 Order      61  metabolism_KO -0.135039 0.130685 0.305676  0.458513  ns
 Order      61     defense_KO  0.039563 0.137421 0.774436  0.774436  ns

--- mean_n_envs ---
 level  n_taxa      predicto